webdictionary/src/main/java/com/eis0/kademlianetwork/informationdeliverymanager/IdFinderHandler.java

```diff
@@ -0,0 +1,163 @@
+        //Create a list containing the closest nodes (five is more than enough)
+        List<SMSKademliaNode> nodeList = table.findClosest(idToFind, ATTEMPTS);
+
+        searchIdList(idToFind, searcher, nodeList);
+    }
+
+    /**
+     * Searches inside of the given list for the active closest node; when a node doesn't respond it
+     * gets removed from the list, and the process is repeated
+     *
+     * @param idToFind The ID to find in the network
+     * @param searcher The peer which is searching for the ID
+     * @param nodeList
+     */
+    public static void searchIdList(@NonNull KademliaId idToFind, @NonNull SMSPeer searcher, @NonNull List<SMSKademliaNode> nodeList) {
+        if (nodeList.isEmpty()) throw new IllegalArgumentException(NODES_LIST_EMPTY);
+        if (nodeList == null) throw new NullPointerException(NODES_LIST_NULL);
+
+        //Local node, inserted in the list if not already present
+        SMSKademliaNode local = KademliaJoinableNetwork.getInstance().getLocalNode();
+        if (!nodeList.contains(local)) {
+            //inserted in the last position, must be the last node to be checked
+            nodeList.add(local);
+        }
+        //Get the node with the node ID closest to the idToFind
+        //The first node in the nodeList is the one with the node Id closest to the idToFind
+        SMSKademliaNode closestNode = nodeList.get(0);
+
+        // Checks if I'm closest node to the one to find
+        if (closestNode.equals(KademliaJoinableNetwork.getInstance().getLocalNode())) {
+            sendResult(idToFind, searcher);
+            return;
+        }
+        //else, I ask to the closest node inside my Routing Table to continue the research
+        SMSPeer closer = closestNode.getPeer();
+        keepLooking(idToFind, searcher, closer);
+        retryIfDead(idToFind, searcher, closer, nodeList);
+    }
+
+    /**
+     * This method sends to the specified target {@link SMSPeer} the result of the previously
+     * carried out research
+     *
+     * @param idToFind   The ID whose research originated the request;
+     *                   It's sent back as response to the research
+     * @param targetPeer The Peer representing the target that will receive the result
+     */
+    private static void sendResult(KademliaId idToFind, SMSPeer targetPeer) {
+        SMSMessage searchResult = new KademliaMessage()
+                .setPeer(targetPeer)
+                .setRequestType(RequestTypes.FindIdSearchResult)
+                .setIdToFind(idToFind)
+                .buildMessage();
+        if (targetPeer.equals(KademliaJoinableNetwork.getInstance().getLocalNode().getPeer())) {
+            //If I'm searching the id for myself, then I have a pending request I can directly fulfill
+            KademliaJoinableNetwork.getInstance().getRequestsHandler().completeFindIdRequest(idToFind, targetPeer);
```

Comment:
Every action is started by a Request that remains opened since its completion; if the node which started the request is the targetPeer, then we can avoid communicating with an SMS message, locally completing the Request.